### ISSUE TITLE
loss=nan issue in Tensorflow 2

### ISSUE THREAD

**Issue by jshu12** [tensorflow/tensorflow#52053]
There is a bug in Tensorflow 2 that happens when all of the following conditions are met:
1.	Multi-GPU is enabled. 
2.	Custom loss function is used. 
3.	The number of training samples is not an integer multiple of the batch size (e.g. 3 training samples and a batch size of 2). 

Under these conditions, you'll end up with a nan loss. 
(The bug also happens if you use multiple GPU cards with batch_size=1, nobody would actually do that.)
I have tried Tensorflow 2.1, 2.4, and 2.5, so it seems like the bug is in all versions of Tensorflow 2. 

Here is a sample code you can run to reproduce the issue:

```
import os
os.environ['CUDA_VISIBLE_DEVICES'] = "0,1" # use a single GPU card, and bug doesn't happen
import tensorflow as tf
physical_devices = tf.config.list_physical_devices('GPU')
for device in physical_devices:
  tf.config.experimental.set_memory_growth(device, True)
print("tensorflow.__version__: " + str(tf.__version__))
import numpy as np
Model = tf.keras.models.Model
Input = tf.keras.layers.Input
Conv2D = tf.keras.layers.Conv2D
K = tf.keras.backend
strategy = tf.distribute.MirroredStrategy()

num_training_samples = 3 # if = 4, bug doesn't happen because 4 is a multiple of 2
batch_size = 2 # bug also happens with batch_size = 1 if you set multiple GPU cards above, regardless of if the number of training samples is an even multiple of the number of GPU cards

inputdata = np.zeros((num_training_samples,5,5,1))
outputdata = np.ones((num_training_samples,5,5,1))

with strategy.scope(): # remove this, and bug doesn't happen
  input_layer = Input(shape=(5,5,1))
  output_layer = Conv2D(filters=1, kernel_size=(3,3), padding='same', activation='tanh', data_format='channels_last', kernel_initializer='glorot_uniform')(input_layer)
  model = Model(inputs=input_layer, outputs=output_layer)

  def custom_loss(y_true, y_pred):
    diff = K.square(y_true - y_pred)
    loss = K.mean(diff)
    return loss

  model.compile(loss=custom_loss, # if loss='mse', bug doesn't happen
                optimizer='sgd')

options = tf.data.Options()
options.experimental_distribute.auto_shard_policy = tf.data.experimental.AutoShardPolicy.OFF
train_dataset = tf.data.Dataset.from_tensor_slices((inputdata, outputdata))
train_dataset = train_dataset.with_options(options)
train_dataset = train_dataset.batch(batch_size)
model.fit(train_dataset, epochs=10, verbose=1)
#model.fit(inputdata, outputdata, batch_size=batch_size, epochs=10, verbose=1) # bug still happens even if you don't use Tensorflow Dataset
```

**Comment by sanatmpa1** [comment]
@jvishnuvardhan,

As colab doesn't support multi GPU environment, The code runs fine([gist](https://colab.research.google.com/gist/sanatmpa1/face09972613ddec08460a1a397358db/52053.ipynb)) and I am not able to replicate the issue. Can you take a look into this one?

**Comment by bguetarni** [comment]
This could be due to the lack of precision about the reduction strategy used. Could you please look at [this](https://www.tensorflow.org/tutorials/distribute/custom_training#define_the_loss_function) and tell us if it helps.

**Comment by rishishounak** [comment]
I am having a similar issue getting loss nan when using gpu but training works fine when using cpu

**Comment by jvishnuvardhan** [comment]
@jshu12 Can you please check with recent TF versions and let us know whether this is still an issue for you. Meanwhile, I would also check on my system. Thanks!

**Comment by jshu12** [comment]
It takes a lot of effort for my team to install a new version of Tensorflow into our tool, and this bug doesn't justify that effort. So I won't be able to verify on a newer version of Tensorflow myself. Especially because we worked around the bug by padding with extra empty training samples until the number of training samples is a multiple of batch size.

**Comment by jshu12** [comment]
I verified on Tensorflow 2.8 and the bug is still there.

**Comment by gadagashwini** [comment]
I could able to replicate with Tensorflow 2.9.1.
<img width="1565" alt="Screen Shot 2022-07-25 at 5 24 16 PM" src="https://user-images.githubusercontent.com/99852755/180775176-8983f7b8-82e1-4558-bde0-8a7dc9345b36.png">

**Comment by fhijo15** [comment]
I still have the same issue in tf 2.10

**Comment by fhijo15** [comment]
Is there any way to fix it ?

**Comment by fhijo15** [comment]
Or any custom training loop that can solve this issue without using keras model.fit ?

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

I tried to execute the code on tensorflow v2.17 which contains Keras3.0 and the code was executed without fail. Kindly find the gist  of it [here](https://colab.research.google.com/gist/tilakrayal/edc202c47225064af4ada5ace8adf35c/untitled2164.ipynb).

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
